update shipment-provider-api (#593)
bigcommerce/dev-docs: docs/api-docs/store-management/shipping/shipping-provider-api.md

<div class="otp" id="no-index">

### On this page
- [Sign-up](#sign-up)
- [Before development](#before-development)
- [Developing the app](#developing-the-app)
- [Control panel installation workflow](#control-panel-installation-workflow)
- [Validation credentials](#validation-credentials)
- [API installation workflow](#api-installation-workflow)
- [Returning shipping quotes](#returning-shipping-quotes)
- [Including product metatadata in rate requests](#including-product-metatadata-in-rate-requests)
- [Submitting the app](#submitting-the-app)
- [App diagram](#app-diagram)
- [Definitions](#definitions)
- [FAQ](#faq)
- [Resources](#resources)

</div>

> Shipping service providers wishing to offer shipping services and rates to BigCommerce merchants and shoppers can implement BigCommerce Shipping Provider endpoints. Once they implement and accept their service into BigCommerce's shipping Carrier Registry, merchants will have access to enable and configure the service through their BigCommerce control panel. Once enabled on a store, BigCommerce will automatically retrieve the service options and rates via the provider's endpoints and display them to merchants in the store's control panel and to shoppers on the storefront.

Reviewer: "via" should not be used per the Style Guide Word List.
```suggestion
Shipping service providers wishing to offer shipping services and rates to BigCommerce merchants and shoppers can implement BigCommerce Shipping Provider endpoints. Once they implement and accept their service into BigCommerce's shipping Carrier Registry, merchants will have access to enable and configure the service through their BigCommerce control panel. Once enabled on a store, BigCommerce will automatically retrieve the service options and rates using the provider's endpoints and display them to merchants in the store's control panel and to shoppers on the storefront.

```

* Email
* Description of the app

> To get your app ID, create an app in [Developer Tools](https://devtools.bigcommerce.com/), and fill out the information on [Step 3 Technical](https://developer.bigcommerce.com/api-docs/apps/guide/requirements#technical). In the URL, the app will have a unique ID. Send the unique ID in exchange for a carrier ID to test the app.

Reviewer: The link for Step 3 Technical takes you to App Store Approval Requirements.

Author: I changed the link. 
I think this link will work.
https://developer.bigcommerce.com/api-docs/apps/guide/publishing#add-technical-information

![#### App ID](//s3.amazonaws.com/user-content.stoplight.io/6012/1552664114224 "#### App ID")

## Before development

BigCommerce will send requests to your server to get information back about shipping quotes, credential validation, and configuration.

### Your service URLs
Since BigCommerce will be sending requests, you will need to provide BigCommerce with URLs that can accept quote requests and optionally a URL to check and validate connection options during app registration. These can be any valid HTTPS URL using port `443`, for example:

`https://yourhost.com/rate`

Replace your host and rate with your own host and path.

### Routes

You should create a URL to provide shipping quotes on your API and check the available shipping rates. The second URL (optional) checks to ensure the merchant’s connection settings are valid. It can perform any checks necessary to do so, such as looking up credentials in your database or calling a downstream service to verify them.

### Requests and responses

> BigCommerce will send and receive data using JSON. The request for rates will always be formatted using the [Base Rate Request Model](https://developer.bigcommerce.com/api-reference/store-management/shipping-provider-api/shipping-provider/requestshippingrates). The response for rates should be formatted using the [Carrier Quote Object](https://developer.bigcommerce.com/api-reference/store-management/shipping-provider-api/shipping-provider/requestshippingrates). Format the request to check for merchant app credentials as Check Connection Options Request Payload, and the response should be formatted using the Check Connection Options Response Payload.

Reviewer: The Base Rate Request Model link takes you to the Request Shipping Rates endpoint.

> BigCommerce will send and receive data using JSON. The request for rates will always be formatted using the [Base Rate Request Model](https://developer.bigcommerce.com/api-reference/store-management/shipping-provider-api/shipping-provider/requestshippingrates). The response for rates should be formatted using the [Carrier Quote Object](https://developer.bigcommerce.com/api-reference/store-management/shipping-provider-api/shipping-provider/requestshippingrates). Format the request to check for merchant app credentials as Check Connection Options Request Payload, and the response should be formatted using the Check Connection Options Response Payload.

Reviewer: ```suggestion
BigCommerce will send and receive data using JSON. The request for rates will always be formatted using the [Base Rate Request Model](https://developer.bigcommerce.com/api-reference/store-management/shipping-provider-api/shipping-provider/requestshippingrates). The response for rates should be formatted using the [Carrier Quote Object](https://developer.bigcommerce.com/api-reference/store-management/shipping-provider-api/shipping-provider/requestshippingrates). Format the request to check for merchant app credentials as Check Connection Options request payload, and the response should be formatted using the Check Connection Options response payload.

```

}
```

## Developing the app

The intended use of the Shipping Provider API is to create an app that merchants can install on their store. The Shipping Provider API can be a standalone app or part of an existing application. When developing the app, there are a few things to consider, which are listed below.

## Control Panel installation workflow

During the app setup, if you configure the Check Connection Options URL for the carrier, an attempt to connect the carrier via the Shipping Manager UI or the Connect Carrier API causes a request to be made to that URL with the provided options. The resource should respond by indicating if the credentials are valid and explain what is wrong. If you did not configure this URL, this check is not required and the credentials are assumed valid as long as they pass type checks.

> **Example request check connection**

Reviewer: ```suggestion
To validate connection options, send a `POST` request to `http://<yourserver.com>/check_connection_options`:

```

> `/POST https://developerserver.com/check_connection_options`

Reviewer: I think we can remove this line of code.

}
```

> **Example response check connection**

Reviewer: ```suggestion
Response:

```

> `/POST https://developerserver.com/check_connection_options`

Reviewer: I think this `/POST` code can be removed.

<!-- theme:  -->

## Validation credentials
> The step of validating the credentials is optional. It does not change how the app operates. It is best practice to authenticate the user against your database or the downstream provider service.

</div>
</div>
</div>

Once you install the app, it will be made available for configuration by merchants and API users. A merchant can navigate to the Shipping Manager and enable, configure, and disable the carrier for any defined zone.

## API installation workflow

To set up a carrier using the API, first, connect it using the Connect Carrier API. Make a request containing the connection settings required by your carrier. The ID of the carrier is required. The carrier ID will be issued by BigCommerce when your carrier is registered. All connection fields are unique per carrier. If your carrier doesn’t require any connection settings, then this object can be left empty.

<!--
title: "Sample Request "
subtitle: "POST https://developerserver.com/shipping/carrier/connection"
lineNumbers: true
-->

> **Example request carrier connection**

Reviewer: ```suggestion
To create a carrier connection, send a `POST` request. For example:

```

> `/POST https://developerserver.com/shipping/carrier/connection`

Reviewer: I think we can remove this line of code or add a link to the [Create a Carrier Connection](https://developer.bigcommerce.com/api-reference/store-management/shipping-api/shipping-carrier/postshippingcarrierconnection) endpoint.

Author: I will remove the code.

lineNumbers: true
-->

**Example request with empty object**

> `/POST https://developerserver.com/shipping/carrier/connection`

Reviewer: I think we can remove this line of code.

Author: Yes, sounds good.

}
```

Once connected, it’s possible to create shipping methods for a connected carrier in any shipping zone. You can query shipping zones using the Shipping Zones resource. For any zone, a request can be made to the Shipping Methods resource using the zone ID from the Shipping Zones resource to create a new method for the connected carrier. You are required to enter the shipping carrier’s ID in the type field.
<!--
title: "Sample Request"
subtitle: "POST https://api.bigcommerce.com/stores/{store_hash}/v2/shipping/zones/{zone_id}/methods"
lineNumbers: true
-->

> **Example request shipping method**

Reviewer: ```suggestion
To create a shipping method, send a `POST` request. For example:

```

lineNumbers: true
-->

> **Example response shipping methods**

Reviewer: ```suggestion
Response:

```

}
```

## Returning shipping quotes

Whenever shipping rates are required, BigCommerce checks its internal cache for valid entries. BigCommerce uses valid entries and does not call the shipping carrier. If a valid cache entry does not exist, BigCommerce makes a request to the Quote URL with details of the items to be shipped, the shipping origin, and the shipping destination. If you configured any connection settings or zone settings, include these. The shipping carrier must then respond with zero or more shipping quotes.

<!--
title: "Sample Request"
subtitle: "POST https://developerserver.com/rate"
lineNumbers: true
-->

> **Example request shipping rates**

Reviewer: ```suggestion
To request shipping rates, send a `POST` request. For example: 

```

lineNumbers: true
-->

> **Example response shipping rates**

Reviewer: ```suggestion
Response:

```

}
```

When you uninstall an app with an associated shipping carrier, you also automatically remove all of the shipping methods and connection info for that carrier from the store. You can no longer make quote requests, and users will no longer see shipping quotes for that carrier.

## Including product metatadata in rate requests

BigCommerce passes carrier-specific product metadata to a carrier service in a rate request via product and variant metafields. This product metadata can be useful if your service depends on specific fields that are not existent on BigCommerce products or variants by default.

To pass metadata in a rate request, the metafields must meet the following requirements:

* must be a product or variant metafields (you cannot pass category, brand, and other metafields in rate requests)
* `permission_set` on the metafield must be `read` or `write`
* `namespace` on the metafield must match this format: `shipping_carrier_carrier_id` (example: `shipping_carrier_72`)

> The carrier registration process described in the [Sign-up Section](https://developer.bigcommerce.com/api-docs/store-management/shipping/shipping-provider-api#sign-up) provides the `carrier_id`.

Reviewer: ```suggestion
The carrier registration process described in the [Sign-up Section](https://developer.bigcommerce.com/api-docs/providers/shipping#sign-up) provides the `carrier_id`.

```

For more information on product and variant metafields, see:
* [API Reference > Store Management > Catalog > Product Metafields](https://developer.bigcommerce.com/api-reference/catalog/catalog-api/product-metafields)
* [API Reference > Store Management > Catalog > Product Variant Metafields](https://developer.bigcommerce.com/api-reference/catalog/catalog-api/product-variants-metafields)

## Submitting the app

Before submitting your app, make sure you have the following information.

> **Single carrier vs. multi carrier**

Reviewer: ```suggestion
**Single-carrier vs. multi-carrier**

```

> Whether the app is a single carrier or multiple carriers. A single carrier app will offer one service, such as USPS. A multi carrier app will offer more than one carrier such as USPS, DHL, and Canada Post.

Reviewer: ```suggestion
A single-carrier app will offer one service, such as USPS. A multi-carrier app will offer more than one carrier such as USPS, DHL, and Canada Post.

```
